### DISCUSSION TITLE
Implementing the mortar Jacobian

### DISCUSSION THREAD

**Discussion by hugary1995** [idaholab/moose#21771] [Q&A General] [answered]
Hi, I am trying to implement a non-AD version of `PenaltyEqualValueConstraint` as follows:
```cpp
Real
RegularPenaltyEqualValueConstraint::computeQpResidual(Moose::MortarType mortar_type)
{
  switch (mortar_type)
  {
    case Moose::MortarType::Secondary:
    {
      auto residual =
          -(_u_primary[_qp] - _u_secondary[_qp]) * _penalty_value * _test_secondary[_i][_qp];

      return residual;
    }

    case Moose::MortarType::Primary:
    {
      auto residual =
          (_u_primary[_qp] - _u_secondary[_qp]) * _penalty_value * _test_primary[_i][_qp];

      return residual;
    }

    default:
      return 0;
  }
}

Real
RegularPenaltyEqualValueConstraint::computeQpJacobian(Moose::ConstraintJacobianType jacobian_type,
                                                      unsigned int jvar)
{
  typedef Moose::ConstraintJacobianType JType;

  switch (jacobian_type)
  {
    case JType::SecondarySecondary:
      if (jvar == _secondary_var.number())
        return (*_phi)[_j][_qp] * _penalty_value * _test_secondary[_i][_qp];
      break;

    case JType::SecondaryPrimary:
      if (jvar == _primary_var.number())
        return -(*_phi)[_j][_qp] * _penalty_value * _test_secondary[_i][_qp];
      break;

    case JType::PrimarySecondary:
      if (jvar == _secondary_var.number())
        return -(*_phi)[_j][_qp] * _penalty_value * _test_primary[_i][_qp];
      break;

    case JType::PrimaryPrimary:
      if (jvar == _primary_var.number())
        return (*_phi)[_j][_qp] * _penalty_value * _test_primary[_i][_qp];
      break;

    default:
      return 0;
  }

  return 0;
}
```

However, the matrix assembled using the non-AD mortar constraint isn't the same as the AD version (it is mainly missing some off-diagonal entries). I must be missing some Jacobian contributions somewhere. @lindsayad @recuero do you have any idea what might be going on here?

Thanks,
Gary

**Comment by hugary1995** [comment]
I also commented out line 145-147 in `MortarConstraint.C`:
```cpp
      // There's no actual coupling between secondary and primary dofs
      // if ((jacobian_type == JType::SecondaryPrimary) || (jacobian_type ==
      // JType::PrimarySecondary))
      //   continue;
```

**Comment by ttruster** [comment]
All:

I've also noticed a similar issue when deriving from the MortarConstraint class that the assembled matrix seems to miss the values that are being computed for PrimarySecondary and SecondaryPrimary.
I'll add that I have also commented the lines
      // There's no actual coupling between secondary and primary dofs 
      if ((jacobian_type == JType::SecondaryPrimary) || (jacobian_type == JType::PrimarySecondary))
        continue;
within the MortarConstraint.C file in the 7/29/2022 version of the code. And I can confirm from VSCode debugging that there are some non-zero entries in the local k matrix of the PrimarySecondary, and a call is made to accumulate that matrix.

Thanks,

Tim

**Comment by lindsayad** [comment, marked as answer]
Are you guys making sure that you are using an SMP matrix with `full = true`? If you use `solve_type = NEWTON` or you have AD objects in your simulation, then we programmatically set that in the framework. But if you are using PJFNK with no AD objects, then by default you will only get diagonal matrix entries

**Reply by lindsayad** [reply]
although if you are running this with the same variable for both `secondary_variable` and `primary_variable` then that is not a concern

**Reply by hugary1995** [reply]
I was using newton, and the same variable on primary and secondary.

**Reply by lindsayad** [reply]
I see the problem. `Assembly::cacheJacobianMortar` does not accumulate the `SecondaryPrimary` or `PrimarySecondary` blocks. That method is unused by AD but is used by the hand-coded objects. Do you think you'll put up a PR for this non-AD object? If so, you (or I) could put the fix for that method in that PR

**Reply by hugary1995** [reply]
I see! I'll do it. Thanks!
